### DISCUSSION TITLE
Example of application custom Theme

### DISCUSSION THREAD

**Discussion by sum-elier** [hecrj/iced#1431] [Q&A] [answered]
I can't in find an example using a custom [`Theme in trait Application`](https://github.com/iced-rs/iced/blob/ab1ac8228f2ab9ef53303e2c5b1770eaa9a513ad/src/application.rs#L105-L108) (using `HEAD` version of Iced).
Defining a custom theme generates the following error:

```rust
error[E0053]: method `view` has an incompatible type for trait
   --> main_app.rs:217:26
    |
217 |    fn view(&mut self) -> Element<'_, GlobalMsg> {
    |                          ^^^^^^^^^^^^^^^^^^^^^^
    |                          |
    |                          expected struct `MainTheme`, found enum `Theme`
    |                          help: change the output type to match the trait: `iced_native::Element<'_, crtapp::GlobalMsg, iced_graphics::renderer::Renderer<iced_wgpu::backend::Backend, MainTheme>>`
    |
    = note: expected fn pointer `fn(&mut Crt01App) -> iced_native::Element<'_, _, iced_graphics::renderer::Renderer<_, MainTheme>>`
               found fn pointer `fn(&mut Crt01App) -> iced_native::Element<'_, _, iced_graphics::renderer::Renderer<_, Theme>>`
```

And the relevant excerpt of the trait implementation:

```rust
impl Application for Crt01App {
   type Theme = MainTheme;
   type Message = GlobalMsg;
   ...

   fn view(&mut self) -> Element<'_, GlobalMsg> {
	...
   }
}
```

All the examples use Iced's [`Theme`](https://github.com/iced-rs/iced/blob/ab1ac8228f2ab9ef53303e2c5b1770eaa9a513ad/style/src/theme.rs#L22-L28) thus no guidance on how to use a custom theme from scratch. Could there be an example in Iced making use of a theme as described before? Is the `Theme` associated type intended to be used like shown before?

**Comment by discordance** [comment]
I would be interested to know how to use a Custom theme too, working with the integration example: Application trait is not used, only `Renderer` and `program::State::new`

**Comment by hecrj** [comment, marked as answer]
You will need to specify the `Renderer` generic type in `Element`, which is `iced::Renderer<iced::Theme>` by default:

```rust
impl Application for Crt01App {
   type Theme = MainTheme;
   type Message = GlobalMsg;
   ...

   fn view(&self) -> Element<'_, GlobalMsg, iced::Renderer<Self::Theme>> {
	...
   }
}
```

**Reply by discordance** [reply]
I have the issue that the associated type `Theme` is not a member of trait `Program`
not a member of trait `Program`. How do we work around in Apps that don't impl Application trait ?

Thx !

**Comment by sum-elier** [comment]
Great, @hecrj! Nice answer, it helped me solve the issue with the `Application` trait implementation.
Now, related errors still plague my code whenever I want to use the custom theme with the widgets.
Here's a simplification of how I am defining `MainTheme`, copying how Iced defined its own Theme, and using it with the widgets:

```rust
#[derive(Default)]
pub struct MainTheme;

impl application::StyleSheet for MainTheme {
   type Style = MainTheme;

   fn appearance(&self, _style: Self::Style) -> application::Appearance {
      application::Appearance {
         text_color: Color::BLACK,
         // more definitions here
      }
   }
}

#[derive(Debug, Default, Clone, Copy)]
pub enum ActionBtn {
   #[default]
   Open,
   // more buttons styles
}

impl button::StyleSheet for MainTheme {
   type Style = ActionBtn;

   fn active(&self, style: Self::Style) -> button::Appearance {
      match style {
         Self::Style::Open => {
            button::Appearance {
               text_color: Color::WHITE,
               // more definitions here
            }
         }
      }
   }

   fn hovered(&self, style: Self::Style) -> button::Appearance {
      ...
   }

   fn pressed(&self, style: Self::Style) -> button::Appearance {
      ...
   }
}
```

And the error is:
```rust
error[E0308]: mismatched types
   --> main_app.rs:433:14
    |
433 |       .style(styling::ActionBtn::Open)
    |        ----- ^^^^^^^^^^^^^^^^^^^^^^^^ expected enum `iced::theme::Button`, found enum `ActionBtn`
    |        |
    |        arguments to this function are incorrect
    |
note: associated function defined here
   // more output here
```

I could solve this issue by making explicit the generic parameter `<MainTheme>` wherever the button is instantiated/returned,
but I have the sense I am not properly implementing the theme as I end up with new errors related to `iced_native::text::Renderer`.
What's the proper way to define a custom theme?

**Reply by 0x192** [reply]
Your `MainTheme` definition seems right to me. :man_shrugging: 

I've successfully implemented several custom themes in my software and I'm pretty happy with the current result. Maybe it can help you :

- [`theme.rs`](https://github.com/0x192/universal-android-debloater/blob/main/src/core/theme.rs) 
- [`style.rs`](https://github.com/0x192/universal-android-debloater/blob/main/src/gui/style.rs)

**Reply by sum-elier** [reply]
@0x192 This did the trick. Appreciated!

**Comment by discordance** [comment]
Ok, I've managed to fix it on my own.

Just some notes here:

If you don't implement `Application` trait (e.g you are based on the integration_wgpu example, and implementing `Program` instead), here are the steps:

1. Create your own custom theme by duplicating `iced`'s one in `iced_style` crate [here](https://github.com/iced-rs/iced/blob/master/style/src/theme.rs)
2. Rename `Theme` enum to your custom name, e.g `MyTheme`
3. When integrating in the winit event loop, on `state.update`, be sure to pass your custom theme.
```
let _ = state.update(
    viewport.logical_size(),
    conversion::cursor_position(cursor_position, viewport.scale_factor()),
    &mut renderer,
    &MyTheme::Dark, // <----- here
    &renderer::Style {
        text_color: Color::WHITE,
    },
    &mut clipboard,
    &mut debug,
);
```
4. When implementing `Program` trait, be sure to set the associated type Renderer with your custom theme generic param:

```
impl Program for MyApp {
    type Renderer = Renderer<MyTheme>;
    type Message = Message;
    ...
```

5. All your view methods implem should return `iced::Element<Message, Renderer<MyTheme>>`
6. Be sure that everything returned in your view (e.g functions that return view partials etc ...) have also `Renderer<MyTheme>` as generic.
7. Your custom theme must implement all the StyleSheet traits of your custom widgets if any.

A lot of complexity here, but I could solve it like that (was porting a `iced` 0.4 app). I really really hope all this can be simplified or better documented in the future !

Now time to enjoy the new theming API finally !

**Reply by sum-elier** [reply]
I think this should be in its own discussion as the current one concerns explicitly the `Application` trait, otherwise it's going to confuse future readers, and the steps you have delineated won't be easily found for people interested in non `Application` solutions.

**Reply by discordance** [reply]
Fair! I think that steps from 5 to 7 still apply in both senarios though.
